File: src/weaponsprocurement/gui/StockReviewQuoteBook.java
package weaponsprocurement.gui;

import weaponsprocurement.core.SubmarketWeaponStock;
import weaponsprocurement.core.WeaponStockRecord;
import weaponsprocurement.core.WeaponStockSnapshot;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

final class StockReviewQuoteBook {
    static final int PRICE_UNAVAILABLE = Integer.MIN_VALUE;

    private final WeaponStockSnapshot snapshot;
    private final Map<String, List<SubmarketWeaponStock>> sortedBuyStocksByItem = new HashMap<String, List<SubmarketWeaponStock>>();
    private final Map<String, Integer> sellUnitPriceByItem = new HashMap<String, Integer>();
    private final Map<String, Float> unitCargoSpaceByItem = new HashMap<String, Float>();
    private final Map<String, StockReviewQuote> quotesByLine = new HashMap<String, StockReviewQuote>();
    private Map<String, Integer> playerSellUnitPrices;

    StockReviewQuoteBook(WeaponStockSnapshot snapshot) {
        this.snapshot = snapshot;
    }

    List<StockReviewSellerAllocation> sellerAllocations(StockReviewPendingPurchase purchase) {
        return quote(purchase).getSellerAllocations();
    }

    StockReviewPortfolioQuote quotePortfolio(List<StockReviewPendingPurchase> pendingPurchases) {
        StockReviewPortfolioQuote result = new StockReviewPortfolioQuote();
        if (pendingPurchases == null || pendingPurchases.isEmpty()) {
            return result;
        }
        Map<String, Integer> remainingBySource = new HashMap<String, Integer>();
        List<StockReviewPendingPurchase> ordered = StockReviewTradePlanner.executionOrder(pendingPurchases);
        for (int i = 0; i < ordered.size(); i++) {
            StockReviewPendingPurchase purchase = ordered.get(i);
            StockReviewQuote quote = purchase.isBuy()
                    ? quoteBuyWithRemaining(purchase, remainingBySource)
                    : quote(purchase);
            result.addLine(purchase, quote);
        }
        return result;
    }

    int cheapestUnitPrice(WeaponStockRecord record) {
        List<SubmarketWeaponStock> stocks = sortedBuyStocks(record);
        if (stocks.isEmpty()) {
            return Integer.MAX_VALUE;
        }
        return stocks.get(0).getUnitPrice();
    }

    int nextBuyUnitPriceAfterPlannedBuys(WeaponStockRecord record, int plannedBuyQuantity) {
        int planned = Math.max(0, plannedBuyQuantity);
        List<SubmarketWeaponStock> stocks = sortedBuyStocks(record);
        for (int i = 0; i < stocks.size(); i++) {
            SubmarketWeaponStock stock = stocks.get(i);
            if (planned >= stock.getCount()) {
                planned -= stock.getCount();
                continue;
            }
            return stock.getUnitPrice();
        }
        return Integer.MAX_VALUE;
    }

    private StockReviewQuote quote(StockReviewPendingPurchase purchase) {
        if (purchase == null || purchase.isZero()) {
            return StockReviewQuote.ZERO;
        }
        String key = lineKey(purchase);
        StockReviewQuote cached = quotesByLine.get(key);
        if (cached != null) {
            return cached;
        }
        StockReviewQuote result = purchase.isSell() ? quoteSell(purchase) : quoteBuy(purchase);
        quotesByLine.put(key, result);
        return result;
    }

    private StockReviewQuote quoteSell(StockReviewPendingPurchase purchase) {
        int unitPrice = sellUnitPrice(purchase.getItemKey());
        if (unitPrice < 0) {
            return StockReviewQuote.priceUnavailable();
        }
        float cargo = purchase.getQuantity() * fallbackUnitCargoSpace(purchase.getItemKey());
        return new StockReviewQuote(purchase.getQuantity() * unitPrice, cargo,
                Collections.<StockReviewSellerAllocation>emptyList());
    }

    private StockReviewQuote quoteBuy(StockReviewPendingPurchase purchase) {
        return quoteBuyWithRemaining(purchase, null);
    }

    private StockReviewQuote quoteBuyWithRemaining(StockReviewPendingPurchase purchase,
                                                   Map<String, Integer> remainingBySource) {
        int remaining = purchase.getQuantity();
        int totalCost = 0;
        int totalBaseCost = 0;
        int totalQuantity = 0;
        float totalCargo = 0f;
        List<StockReviewSellerAllocation> allocations =
                new ArrayList<StockReviewSellerAllocation>();
        List<SubmarketWeaponStock> stocks = sortedBuyStocks(purchase.getItemKey());
        for (int i = 0; i < stocks.size() && remaining > 0; i++) {
            SubmarketWeaponStock stock = stocks.get(i);
            if (purchase.getSubmarketId() != null && !matchesSource(purchase.getSubmarketId(), stock)) {
                continue;
            }
            int available = remainingBySource == null ? stock.getCount() : remainingStock(purchase.getItemKey(), stock, remainingBySource);
            int quantity = Math.min(remaining, available);
            if (quantity <= 0) {
                continue;
            }
            int cost = quantity * stock.getUnitPrice();
            totalCost += cost;
            totalBaseCost += quantity * stock.getBaseUnitPrice();
            totalQuantity += quantity;
            totalCargo += quantity * stock.getUnitCargoSpace();
            allocations.add(new StockReviewSellerAllocation(stock.getDisplaySourceName(),
                    stock.getSourceId(), quantity, cost));
            remaining -= quantity;
            if (remainingBySource != null) {
                remainingBySource.put(sourceKey(purchase.getItemKey(), stock), Integer.valueOf(available - quantity));
            }
        }
        if (remaining > 0) {
            return StockReviewQuote.priceUnavailable(totalCargo, allocations);
        }
        return new StockReviewQuote(totalCost, totalCargo, totalBaseCost, totalQuantity, allocations);
    }

    private int remainingStock(String itemKey, SubmarketWeaponStock stock, Map<String, Integer> remainingBySource) {
        String key = sourceKey(itemKey, stock);
        Integer cached = remainingBySource.get(key);
        if (cached != null) {
            return cached.intValue();
        }
        remainingBySource.put(key, Integer.valueOf(stock.getCount()));
        return stock.getCount();
    }

    private List<SubmarketWeaponStock> sortedBuyStocks(String itemKey) {
        if (itemKey == null) {
            return Collections.emptyList();
        }
        List<SubmarketWeaponStock> cached = sortedBuyStocksByItem.get(itemKey);
        if (cached != null) {
            return cached;
        }
        WeaponStockRecord record = findRecord(itemKey);
        if (record == null) {
            sortedBuyStocksByItem.put(itemKey, Collections.<SubmarketWeaponStock>emptyList());
            return Collections.emptyList();
        }
        List<SubmarketWeaponStock> result = new ArrayList<SubmarketWeaponStock>();
        for (int i = 0; i < record.getSubmarketStocks().size(); i++) {
            SubmarketWeaponStock stock = record.getSubmarketStocks().get(i);
            if (stock.isPurchasable() && stock.getCount() > 0) {
                result.add(stock);
            }
        }
        sortByPrice(result);
        result = Collections.unmodifiableList(result);
        sortedBuyStocksByItem.put(itemKey, result);
        return result;
    }

    private List<SubmarketWeaponStock> sortedBuyStocks(WeaponStockRecord record) {
        if (record == null) {
            return Collections.emptyList();
        }
        String itemKey = record.getItemKey();
        List<SubmarketWeaponStock> cached = sortedBuyStocksByItem.get(itemKey);
        if (cached != null) {
            return cached;
        }
        List<SubmarketWeaponStock> result = new ArrayList<SubmarketWeaponStock>();
        for (int i = 0; i < record.getSubmarketStocks().size(); i++) {
            SubmarketWeaponStock stock = record.getSubmarketStocks().get(i);
            if (stock.isPurchasable() && stock.getCount() > 0) {
                result.add(stock);
            }
        }
        sortByPrice(result);
        result = Collections.unmodifiableList(result);
        sortedBuyStocksByItem.put(itemKey, result);
        return result;
    }

    private float fallbackUnitCargoSpace(String itemKey) {
        if (itemKey == null) {
            return 1f;
        }
        Float cached = unitCargoSpaceByItem.get(itemKey);
        if (cached != null) {
            return cached.floatValue();
        }
        float result = 1f;
        WeaponStockRecord record = findRecord(itemKey);
        if (record != null) {
            for (int i = 0; i < record.getSubmarketStocks().size(); i++) {
                SubmarketWeaponStock stock = record.getSubmarketStocks().get(i);
                if (stock.getUnitCargoSpace() > 0f) {
                    result = stock.getUnitCargoSpace();
                    break;
                }
            }
        }
        unitCargoSpaceByItem.put(itemKey, Float.valueOf(result));
        return result;
    }

    int sellUnitPrice(String itemKey) {
        if (itemKey == null) {
            return -1;
        }
        Integer cached = sellUnitPriceByItem.get(itemKey);
        if (cached != null) {
            return cached.intValue();
        }
        if (playerSellUnitPrices == null) {
            playerSellUnitPrices = StockReviewPlayerCargo.sellUnitPricesByItem(
                    snapshot == null ? null : snapshot.getMarket(),
                    snapshot != null && snapshot.isIncludeBlackMarket());
        }
        Integer price = playerSellUnitPrices.get(itemKey);
        int result = price == null ? -1 : price.intValue();
        sellUnitPriceByItem.put(itemKey, Integer.valueOf(result));
        return result;
    }

    private WeaponStockRecord findRecord(String itemKey) {
        if (snapshot == null || itemKey == null) {
            return null;
        }
        return snapshot.getRecord(itemKey);
    }

    private static void sortByPrice(List<SubmarketWeaponStock> stocks) {
        for (int i = 0; i < stocks.size(); i++) {
            for (int j = i + 1; j < stocks.size(); j++) {
                if (compareSourceOrder(stocks.get(j), stocks.get(i)) < 0) {
                    SubmarketWeaponStock temp = stocks.get(i);
                    stocks.set(i, stocks.get(j));
                    stocks.set(j, temp);
                }
            }
        }
    }

    private static int compareSourceOrder(SubmarketWeaponStock left, SubmarketWeaponStock right) {
        int result = Integer.compare(left.getUnitPrice(), right.getUnitPrice());
        if (result != 0) {
            return result;
        }
        return left.getDisplaySourceName().compareToIgnoreCase(right.getDisplaySourceName());
    }

    private static String lineKey(StockReviewPendingPurchase purchase) {
        return (purchase.getItemKey() == null ? "" : purchase.getItemKey())
                + "|" + (purchase.getSubmarketId() == null ? "" : purchase.getSubmarketId())
                + "|" + purchase.getQuantity();
    }

    private static String sourceKey(String itemKey, SubmarketWeaponStock stock) {
        return (itemKey == null ? "" : itemKey)
                + "|" + (stock == null || stock.getSourceId() == null ? "" : stock.getSourceId());
    }

    private static boolean matchesSource(String requestedSourceId, SubmarketWeaponStock stock) {
        if (stock == null || requestedSourceId == null) {
            return false;
        }
        return requestedSourceId.equals(stock.getSourceId()) || requestedSourceId.equals(stock.getSubmarketId());
    }

}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/StockReviewQuoteBook.kt
package weaponsprocurement.ui.stockreview

import weaponsprocurement.ui.*

import weaponsprocurement.stock.SubmarketWeaponStock
import weaponsprocurement.trade.TradeMoney
import weaponsprocurement.stock.WeaponStockRecord
import weaponsprocurement.stock.WeaponStockSnapshot
import java.util.ArrayList
import java.util.Collections
import java.util.Comparator
import java.util.HashMap

class StockReviewQuoteBook(private val snapshot: WeaponStockSnapshot?) {
    private val sortedBuyStocksByItem = HashMap<String, List<SubmarketWeaponStock>>()
    private val sellUnitPriceByItem = HashMap<String, Int>()
    private val unitCargoSpaceByItem = HashMap<String, Float>()
    private val quotesByLine = HashMap<String, StockReviewQuote>()
    private var playerSellUnitPrices: Map<String, Int>? = null

    fun sellerAllocations(trade: StockReviewPendingTrade?): List<StockReviewSellerAllocation> {
        return quote(trade).sellerAllocations
    }

    fun quotePortfolio(pendingTrades: List<StockReviewPendingTrade>?): StockReviewPortfolioQuote {
        val result = StockReviewPortfolioQuote()
        if (pendingTrades == null || pendingTrades.isEmpty()) return result

        val remainingBySource = HashMap<String, Int>()
        val ordered = StockReviewTradePlanner.executionOrder(pendingTrades)
        for (trade in ordered) {
            val quote = if (trade.isBuy()) {
                quoteBuyWithRemaining(trade, remainingBySource)
            } else {
                quote(trade)
            }
            result.addLine(trade, quote)
        }
        return result
    }

    fun cheapestUnitPrice(record: WeaponStockRecord?): Int {
        val stocks = sortedBuyStocks(record)
        if (stocks.isEmpty()) return Int.MAX_VALUE
        return stocks[0].unitPrice
    }

    fun nextBuyUnitPriceAfterPlannedBuys(record: WeaponStockRecord?, plannedBuyQuantity: Int): Int {
        var planned = Math.max(0, plannedBuyQuantity)
        val stocks = sortedBuyStocks(record)
        for (stock in stocks) {
            if (planned >= stock.count) {
                planned -= stock.count
                continue
            }
            return stock.unitPrice
        }
        return Int.MAX_VALUE
    }

    private fun quote(trade: StockReviewPendingTrade?): StockReviewQuote {
        if (trade == null || trade.isZero()) return StockReviewQuote.ZERO
        val key = lineKey(trade)
        val cached = quotesByLine[key]
        if (cached != null) return cached
        val result = if (trade.isSell()) quoteSell(trade) else quoteBuy(trade)
        quotesByLine[key] = result
        return result
    }

    private fun quoteSell(trade: StockReviewPendingTrade): StockReviewQuote {
        val unitPrice = sellUnitPrice(trade.itemKey)
        if (unitPrice < 0) return StockReviewQuote.priceUnavailable()
        if (trade.quantity == Int.MIN_VALUE) return StockReviewQuote.priceUnavailable()

        val sellQuantity = -trade.quantity
        val credits = TradeMoney.lineTotal(unitPrice, sellQuantity)
        if (!TradeMoney.canExecuteCreditMutation(credits)) return StockReviewQuote.priceUnavailable()

        val cargo = -sellQuantity * fallbackUnitCargoSpace(trade.itemKey)
        return StockReviewQuote(-credits, cargo, Collections.emptyList())
    }

    private fun quoteBuy(trade: StockReviewPendingTrade): StockReviewQuote {
        return quoteBuyWithRemaining(trade, null)
    }

    private fun quoteBuyWithRemaining(
        trade: StockReviewPendingTrade,
        remainingBySource: MutableMap<String, Int>?,
    ): StockReviewQuote {
        var remaining = trade.quantity
        var totalCost = 0L
        var totalBaseCost = 0L
        var totalQuantity = 0
        var totalCargo = 0f
        val allocations = ArrayList<StockReviewSellerAllocation>()
        val stocks = sortedBuyStocks(trade.itemKey)

        for (stock in stocks) {
            if (remaining <= 0) break
            if (trade.submarketId != null && !matchesSource(trade.submarketId, stock)) continue

            val available = if (remainingBySource == null) {
                stock.count
            } else {
                remainingStock(trade.itemKey, stock, remainingBySource)
            }
            val quantity = Math.min(remaining, available)
            if (quantity <= 0) continue

            val cost = TradeMoney.lineTotal(stock.unitPrice, quantity)
            totalCost = TradeMoney.safeAdd(totalCost, cost)
            totalBaseCost = TradeMoney.safeAdd(totalBaseCost, TradeMoney.lineTotal(stock.baseUnitPrice, quantity))
            totalQuantity += quantity
            totalCargo += quantity * stock.unitCargoSpace
            allocations.add(StockReviewSellerAllocation(stock.displaySourceName, stock.sourceId, quantity, cost))
            remaining -= quantity
            if (remainingBySource != null) {
                remainingBySource[sourceKey(trade.itemKey, stock)] = available - quantity
            }
        }

        if (remaining > 0) {
            return StockReviewQuote.priceUnavailable(totalCargo, allocations)
        }
        return StockReviewQuote(totalCost, totalCargo, totalBaseCost, totalQuantity, allocations)
    }

    private fun remainingStock(
        itemKey: String?,
        stock: SubmarketWeaponStock,
        remainingBySource: MutableMap<String, Int>,
    ): Int {
        val key = sourceKey(itemKey, stock)
        val cached = remainingBySource[key]
        if (cached != null) return cached
        remainingBySource[key] = stock.count
        return stock.count
    }

    private fun sortedBuyStocks(itemKey: String?): List<SubmarketWeaponStock> {
        if (itemKey == null) return Collections.emptyList()
        val cached = sortedBuyStocksByItem[itemKey]
        if (cached != null) return cached
        val record = findRecord(itemKey)
        if (record == null) {
            sortedBuyStocksByItem[itemKey] = Collections.emptyList()
            return Collections.emptyList()
        }
        return sortedBuyStocks(record)
    }

    private fun sortedBuyStocks(record: WeaponStockRecord?): List<SubmarketWeaponStock> {
        if (record == null) return Collections.emptyList()
        val itemKey = record.itemKey
        val cached = sortedBuyStocksByItem[itemKey]
        if (cached != null) return cached

        var result = ArrayList<SubmarketWeaponStock>()
        for (stock in record.submarketStocks) {
            if (stock.isPurchasable() && stock.count > 0) result.add(stock)
        }
        Collections.sort(result, SubmarketStockPriceComparator.INSTANCE)
        val immutable = Collections.unmodifiableList(result)
        sortedBuyStocksByItem[itemKey] = immutable
        return immutable
    }

    private fun fallbackUnitCargoSpace(itemKey: String?): Float {
        if (itemKey == null) return 1f
        val cached = unitCargoSpaceByItem[itemKey]
        if (cached != null) return cached

        var result = 1f
        val record = findRecord(itemKey)
        if (record != null) {
            for (stock in record.submarketStocks) {
                if (stock.unitCargoSpace > 0f) {
                    result = stock.unitCargoSpace
                    break
                }
            }
        }
        unitCargoSpaceByItem[itemKey] = result
        return result
    }

    fun sellUnitPrice(itemKey: String?): Int {
        if (itemKey == null) return -1
        val cached = sellUnitPriceByItem[itemKey]
        if (cached != null) return cached
        if (playerSellUnitPrices == null) {
            playerSellUnitPrices = StockReviewPlayerCargo.sellUnitPricesByItem(
                snapshot?.getMarket(),
                snapshot != null && snapshot.isIncludeBlackMarket(),
            )
        }
        val price = playerSellUnitPrices?.get(itemKey)
        val result = price ?: -1
        sellUnitPriceByItem[itemKey] = result
        return result
    }

    private fun findRecord(itemKey: String?): WeaponStockRecord? {
        if (snapshot == null || itemKey == null) return null
        return snapshot.getRecord(itemKey)
    }

    private class SubmarketStockPriceComparator private constructor() : Comparator<SubmarketWeaponStock> {
        override fun compare(left: SubmarketWeaponStock, right: SubmarketWeaponStock): Int {
            val result = left.unitPrice.compareTo(right.unitPrice)
            return if (result != 0) {
                result
            } else {
                left.displaySourceName.orEmpty().compareTo(right.displaySourceName.orEmpty(), ignoreCase = true)
            }
        }

        companion object {
            @JvmField
            val INSTANCE: SubmarketStockPriceComparator = SubmarketStockPriceComparator()
        }
    }

    companion object {
        const val PRICE_UNAVAILABLE: Int = Int.MIN_VALUE

        private fun lineKey(trade: StockReviewPendingTrade): String {
            return trade.itemKey + "|" + (trade.submarketId ?: "") + "|" + trade.quantity
        }

        private fun sourceKey(itemKey: String?, stock: SubmarketWeaponStock?): String {
            return (itemKey ?: "") + "|" + (stock?.sourceId ?: "")
        }

        private fun matchesSource(requestedSourceId: String?, stock: SubmarketWeaponStock?): Boolean {
            if (stock == null || requestedSourceId == null) return false
            return requestedSourceId == stock.sourceId || requestedSourceId == stock.submarketId
        }
    }
}
